### ISSUE TITLE
Cleaning up threads

### ISSUE THREAD

**Issue by legate** [ymartineau/peers#12]
I tried running the EventManager sample at http://peers.sourceforge.net/files/html/peers.html. If I make a call, then hang up, even though the main thread completes, the program does not stop because there are still some threads running (TransactionManager, InviteHandler, TransportManager). Is there a way I can stop the remaining threads so the program will exit cleanly?

**Comment by ymartineau** [comment]
hi,

here is what is done in peers-gui, in EventManager to close the remaining
threads when user closes the main window:


                try {
                    userAgent.unregister();
                } catch (Exception e) {
                    logger.error("error while unregistering", e);
                }
                closed = true;
                try {
                    Thread.sleep(3 * RFC3261.TIMER_T1);
                } catch (InterruptedException e) {
                }
                System.exit(0);


I think you can do the same.

The example provided in documentation is just a basic example to start
using peers. All use cases are not considered for this demo in usage in
documentation.

**Comment by johnythefarmer** [comment]
Experienced similiar isssue. Using peers library for testing of VoIP system. When the test is over, I cannot call System.exit. Most of the threads that are still going, are TimerThreads (InviteHandler, TransactionManager and RegisterHandler have their timers scheduled, but never cancelled)

**Comment by johnythefarmer** [comment]
The code in pull request is great, but does not solve the register handler problem. Even though the timer is cancelled in RegisterHandler.unregister, there are still TimerThreads going, when the useragent unregisters. The reason is that when RegisterHandler.successResponseReceived is called, new Timer is created and task scheduled, but the old timer is never cancelled and refference to it is lost. This could be solved by calling:
    if (timer != null) {
        timer.cancel();
    }

right before creating new Timer.

**Comment by paraplexed** [comment]
Good catch. I updated the PR to close the timer when refreshing, if a timer already exists.

**Comment by heruan** [comment]
I had the same issue and this worked for me:

```java
userAgent.close()
```

This will stop the `TransportManager`, the `TransactionManager`, the `InviteHandler` and will stop any RTP session started by the `MediaManager`.
